propulsion-hotel/README.md

```diff
@@ -0,0 +1,131 @@
+   the id of the `GroupCheckout` to which it has been merged into), none of
+   the processing is in reaction to events on the `GuestStay` events
+
+## High level processing structure
+
+- A `GroupCheckout` stream maintains a [Todo List](https://blog.bittacklr.be/the-to-do-list-pattern.html)
+  of stays that the Group has been asked to take over the management of
+- That Todo list is the State for the Process Manager (the relevant pieces of
+  the overall State get mapped to `Flow.Action` by `Flow.nextAction`)
+- Propulsion is used to wire from the subscription to the `GroupCheckout` event
+  category to the `Reactor.Handler`. Whenever one or more events are pending
+  processing for a given stream, the relevant backlog of `events` are passed
+  to the `handler`. Propulsion guarantees that there will only be a single
+  handler invocation in flight at any time for a given stream.
+- While the `StaysSelected` event is the primary event that drives the need for
+  the reactor do carry out activities, we read the full state in order that we
+  don't have to re-visit Stays that we know we have already merged into the
+  `GroupCheckout`. Instead, we load the full state of the Checkout and walk
+  only the Stays that have not yet had either a success or failure outcome
+  recorded for them.
+
+## Reading Your Writes / `OverrideWritePosition`
+
+The process is not necessarily dependent on the Reactor being able to
+immediately see the events notified on the change feed.
+
+Examples:
+1. if event 0 is `StaysSelected`, that will trigger the handler. If the
+   Handler attempts to read `GroupCheckout` state_ but that event has not yet
+   propagated to the node from which the Handler reads_, the State will be
+   `Flow.Action.Ready 0`, and the Handler will immediately yield a
+   `SpanResult.OverrideWritePosition 0`
+
+2. if the read _does_ include the `StaysSelected`, then the State will be
+   `Flow.Action.MergeStays { stays = (6 stayIds) }`. The
+   `GroupCheckoutProcess` (concurrently) visits each of the Stays identified
+   by the current pending `stayIds`. If there are 5 successes and 1 failure,
+   the result is `StaysMerged { stays = [{stayId = ..., residual = ...], ...]}`
+   and `MergesFailed { stays = [ (stayid) ]}`. After writing those two events
+   to the stream, the version has moved from `1` to `3`, resulting in
+   `SpanResult.OverrideWritePosition 3`.
+
+   This implies one of the following two possible outcomes:
+   
+   1. Propulsion notes we are at Version 3 and will
+      discard events 1 and 2 on receipt from the change feed, without
+      even entering the streams buffer (and no further handler invocations
+      take place)
+   2. At the point where Propulsion sees the `OverrideWritePosition`, events
+      and/or 2 have already been read and buffered ready for dispatch. In
+      this case, the events are removed from the buffer immediately (and no
+      further handler invocations take place)
```

Comment:
<pre>     
# Getting started

Depending on the concrete store you choose you'll need some setup.

## Message DB

1. Create a docker-compose file

```yaml
version: '3.7'

services:
  messagedb:
    image: ethangarofolo/message-db
    ports:
      - "5432:5432"
```

2. Create the checkpoints table

```sh
$ docker-compose up -d # starts the message-db
$ dotnet new tool-manifest # So we can install the propulsion tool
$ dotnet tool install Propulsion.Tool --prerelease
$ dotnet propulsion initpg -cs public -cc "Host=localhost; Username=postgres; Password=postgres" # creates the checkpoint table
```

3. Run the reactor
```sh
$ CONNSTR="Host=localhost; Username=message_store; Password=" \
  CPCONNSTR="Host=localhost; Username=postgres; Password=postgres" \
  dotnet run --project Reactor -- \
  --processorname MyProcessor \
  mdb -c $CONNSTR -r $CONNSTR -cp $CPCONNSTR -cs public
```
</pre>

(I can't embed markdown code snippets in a suggestion)

Reply:
Not sure how best to document the dynamo setup (do you even run it locally?)

<pre>
# Dynamo

1. Create a docker-compose file

```yaml
version: '3.7'

services:
  dynamodb-local:
    image: amazon/dynamodb-local
    container_name: dynamodb-local
    hostname: dynamodb-local
    restart: always
    volumes:
      - ./docker-dynamodblocal-data:/home/dynamodblocal/data
    ports:
      - 8000:8000
    command: "-jar DynamoDBLocal.jar -sharedDb -dbPath /home/dynamodblocal/data/"

  dynamodb-admin:
    image: aaronshaf/dynamodb-admin
    ports:
      - "8001:8001"
    environment:
      DYNAMO_ENDPOINT: "http://dynamodb-local:8000"
      AWS_REGION: "us-west-2"
      AWS_ACCESS_KEY_ID: local
      AWS_SECRET_ACCESS_KEY: local
    depends_on:
      - dynamodb-local
```

2. Set up the table, indexer, ....

```sh
install the tool, then run initaws?
```

3. Run all of it
</pre>

Reply:
Thanks, applied with some minor edits